AutoNAT with TCP, distinguishing local and global networks

b-zee opened the pull request:
- feat(node): use TCP as transport
- feat(node): add AutoNAT behaviour to swarm
- feat(node): add local flag

bochaco commented on src/url/version_hash.rs:
```diff
@@ -0,0 +1,105 @@
+/// Version Hash corresponding to the register entry hash where the content is stored
+#[derive(Debug, PartialEq, Serialize, Deserialize, Clone, Default, Copy)]
+pub struct VersionHash {
+    register_entry_hash: EntryHash,
```
Perhaps make this generic and not necessarily for a `Register` entry, e.g. `entry_hash`, I hope eventually we have other native content types which work also with hashes and we can use `VersionHash` in the URLs in the same way.

bochaco commented on src/url/mod.rs:
```diff
@@ -1725,8 +1729,8 @@ mod tests {
     fn test_safeurl_parts() -> Result<()> {
         // These two are equivalent.  ie, the xorurl is the result of nrs.to_xorurl_string()
-        let nrsurl = "safe://my.sub.domain/path/my%20dir/my%20file.txt?this=that&this=other&color=blue&v=5&name=John+Doe#somefragment";
-        let xorurl = "safe://my.sub.hnyydyiixsfrqix9aoqg97jebuzc6748uc8rykhdd5hjrtg5o4xso9jmggbqh/path/my%20dir/my%20file.txt?this=that&this=other&color=blue&v=5&name=John+Doe#somefragment";
```
Perhaps use some default/random hash rather than removing it from the test, I believe this test as well as `test_safeurl_to_string` and `test_safeurl_validate` are aimed at making sure we are parsing/encoding/decoding all parts without missing any part, so removing the version/hash from them wouldn't be desirable?

bochaco reviewed (changes requested):
It looks good to me, I left only a couple of minor things.

bochaco commented:
@grumbach I just realised we'll need a commit marked as breaking change.

b-zee committed:
feat(node): add AutoNAT behaviour to swarm

b-zee committed:
feat(node): use TCP as transport

QUIC transport doesn't seem to work well with the (auto) NAT traversal
in libp2p. This might be reversed in the future, if QUIC is properly
supported in combination with other (NAT) protocols.

b-zee committed:
feat(node): add local flag

With the local flag, we can determine whether we want to add local or global addresses to our routing table. This is important so that we do not mix up two different kinds of networks.

b-zee committed:
feat(node): add AutoNAT peer as server

b-zee committed:
fix(testnet): dial TCP peers instead of QUIC

b-zee committed:
feat(node): add peers only that we dialed to RT

b-zee committed:
feat(node): shut down node in case of private NAT

b-zee committed:
feat(node): change timeout for AutoNAT

b-zee committed:
fix(node): increase autonat timeout

See https://github.com/libp2p/rust-libp2p/issues/3986#issuecomment-1562505650

b-zee committed:
fix(node): locally assume peers are reachable
